Frameworks/CoreText/CTFramesetter.mm

```diff
@@ -19,36 +19,45 @@
 #import <algorithm>
 #import "CoreTextInternal.h"
 #import "CGPathInternal.h"
+#import <CFCppBase.h>
 #import "DWriteWrapper_CoreText.h"
 
-@implementation _CTFramesetter : NSObject
-@end
+struct __CTFramesetter : CoreFoundation::CppBase<__CTFramesetter> {
+    __CTFramesetter(CTTypesetterRef typesetter) : _typesetter(typesetter) {
+    }
+
+    inline CTTypesetterRef Typesetter() const {
+        return _typesetter;
+    }
+
+private:
+    woc::StrongCF<CTTypesetterRef> _typesetter;
+};
 
 /**
  @Status Interoperable
 */
 CTFramesetterRef CTFramesetterCreateWithAttributedString(CFAttributedStringRef string) {
-    _CTFramesetter* ret = [_CTFramesetter alloc];
-    ret->_typesetter.attach(static_cast<_CTTypesetter*>(CTTypesetterCreateWithAttributedString(string)));
-    return static_cast<CTFramesetterRef>(ret);
+    auto typesetter = woc::MakeStrongCF<CTTypesetterRef>(CTTypesetterCreateWithAttributedString(string));
+    return __CTFramesetter::CreateInstance(kCFAllocatorDefault, typesetter);
 }
 
 /**
  @Status Caveat
  @Notes frameAttributes parameter ignored
 */
-CTFrameRef CTFramesetterCreateFrame(CTFramesetterRef framesetterRef, CFRange range, CGPathRef path, CFDictionaryRef frameAttributes) {
-    RETURN_NULL_IF(framesetterRef == nil || path == nullptr);
+CTFrameRef CTFramesetterCreateFrame(CTFramesetterRef framesetter, CFRange range, CGPathRef path, CFDictionaryRef frameAttributes) {
+    RETURN_NULL_IF(framesetter == nil || path == nullptr);
+
     CGRect frameRect = CGPathGetBoundingBox(path);
-    _CTFramesetter* framesetter = static_cast<_CTFramesetter*>(framesetterRef);
 
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
-    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
+    CFAttributedStringRef attributedString = __CTTypesetterGetAttributedString(framesetter->Typesetter());
     if (range.length == 0L) {
-        range.length = [typesetter->_string length] - range.location;
+        range.length = CFStringGetLength(CFAttributedStringGetString(attributedString)) - range.location;
```

Comment:
why not CFAttributedStringGetLength? Technically, an NSAttributedString subclass can override `-length`, and with the way you've implemented this, it will not get called.